### ISSUE TITLE
Resyntax integration

### ISSUE THREAD

**Issue by jackfirth** [jeapostrophe/racket-langserver#40]
In the long term, I would like to integrate [Resyntax](github.com/jackfirth/resyntax/) with this language server so that Racket users can get helpful suggestions for improving their code directly in the editor. I don't think Resyntax is stable enough for this yet, but it's a goal I'm working towards. Does this seem like a good idea to folks?

**Comment by jeapostrophe** [comment]
Absolutely. My hope is that the Racket LSP will provide all of the same kinds of things that other LSPs do and I think Resyntax is the thing in the Racket world that fits the mold.

**Comment by spdegabrielle** [comment]
@jackfirth I think resyntax might be _stable enough_ 😁

**Comment by dannypsnl** [comment]
I make a small exploration
```racket
#lang racket/base
(require resyntax
         resyntax/default-recommendations
         resyntax/private/source
         rebellion/base/range
         rebellion/base/comparator
         rebellion/collection/range-set)

(define all-lines (range-set (unbounded-range #:comparator natural<=>)))

(resyntax-analyze
  (file-source "/Users/dannypsnl/tr-notes/tr/private/next.rkt")
  #:suite default-recommendations
  #:lines all-lines)
```

I think this is almost what LSP expecting, we only need to transform its output structures to LSP structures.

cc @6cdh

**Comment by jackfirth** [comment]
I think you'll also have to output some `Diagnostic` messages too.

**Comment by 6cdh** [comment]
Yes. We need to add a new function that process integration, send diagnostic and code action to [diagnostic.rkt](https://github.com/jeapostrophe/racket-langserver/blob/master/service/diagnostic.rkt).

I also expanded @dannypsnl code a bit, it should print all information we need to implement the integration:

```racket
#lang racket/base

(require resyntax
         resyntax/default-recommendations
         resyntax/private/source
         resyntax/private/refactoring-result
         resyntax/private/string-replacement
         rebellion/base/range
         rebellion/base/comparator
         rebellion/collection/range-set
         racket/string)

(require racket/pretty)

(define all-lines (range-set (unbounded-range #:comparator natural<=>)))

(define result-set
  (resyntax-analyze
    (file-source "or.rkt")
    #:suite default-recommendations
    #:lines all-lines))

(define results (refactoring-result-set-results result-set))

(define (extract-result result)
  (define sr (refactoring-result-string-replacement result))
  (pretty-display sr)
  (define info
    (list
      (format "name: ~a" (refactoring-result-rule-name result))
      (format "kind: refactor")
      (format "message: ~a" (refactoring-result-message result))
      (let ([start (string-replacement-start sr)]
            [end (string-replacement-original-end sr)])
        (format "range: [~a, ~a)" start end))
      (format "newText: ~v" (string-append* (map inserted-string-contents (string-replacement-contents sr))))))

  (displayln (string-join info "\n")))

(for-each extract-result results)
```

I can implement it later.

Also, @jackfirth why `(string-replacement-contents sr)` returns a list of inserted string (not a single inserted string) and they all share the same `start`, `original-span`, `original-end`, ... properties. Is it safe to concat them?

**Comment by jackfirth** [comment]
@6cdh The string replacement doesn't return a single string because if one replacement wants to rewrite `a b c` to `A b C`, and another replacement wants to rewrite `a b c` to `a B c`, then by representing each replacement as a list of individual edits, it becomes possible to combine those two replacements into a single replacement. If the string replacements only contained a single new string inside them, it's be much more difficult to detect when two suggestions are non-overlapping and can be combined together.

This is mostly relevant to the Resyntax Autofixer, which wants to apply multiple suggestions at once to produce [pull requests](https://github.com/racket/scribble/pull/515) with commits that apply many suggestions together. For an editor, I assume we'd only be applying one fix at a time? Or at most, a handful of fixes of the same type in a single file?

**Comment by jackfirth** [comment]
If you want to get just the new text for a string replacement, use the `string-replacement-render` function. Given a string replacement and the full text of the file, that will return a single string representing everything the replacement will insert in its range.

**Comment by 6cdh** [comment]
Thank you. I've updated the code.

**Comment by jryans** [comment]
@6cdh Should this be closed now that #153 has merged, or is there more to do?

**Comment by jackfirth** [comment]
I agree. This is closed, in my view. Future improvements to the integration can be filed as separate feature requests.
